### PR TITLE
feat: Option to throw on error status codes

### PR DESCRIPTION
**Pull request by kibertoad**
refs #1205

### CODE REVIEW

**Comment by ronag** [comment]
I think this is outside of the scope of what undici should be doing.

**Comment by kibertoad** [comment]
@ronag What is the downside? It is a very simple piece of logic that significantly improves developer experience. Reimplementing it for every project is wasteful boilerplate.

**Comment by kibertoad** [comment]
@mcollina Could use your perspective here as well.

File: test/client.js [hunk 1]
```diff
@@ -315,6 +315,64 @@ test('basic get with query params partially in path', (t) => {
...
+      keepAliveTimeout: 300e3
+    })
+    t.teardown(client.close.bind(client))
+
+    const signal = new EE()
+    client.request({
+      signal,
+      path: '/',
+      method: 'GET',
+      throwOnError: true
```
Comment (mcollina) [inline test/client.js:338]: We need a better name for the option, in this example is not throwing at all.

File: types/dispatcher.d.ts [hunk 2]
```diff
@@ -57,6 +57,8 @@ declare namespace Dispatcher {
     headersTimeout?: number | null;
     /** The timeout after which a request will time out, in milliseconds. Monitors time between receiving body data. Use 0 to disable it entirely. Defaults to 30 seconds. */
     bodyTimeout?: number | null;
+    /** Whether Undici should throw an error upon receiving a 4xx or 5xx response from the server. Defaults to false */
+    throwOnError?: boolean;
```
Comment (mcollina) [inline types/dispatcher.d.ts:61]: the additional type needs a test

**Review by mcollina — COMMENTED** [review]
I'm in favor for this change too.

**Comment by ronag** [comment]
I'm on the losing end again 😢 ... can we at least make the error signature match that of https://www.npmjs.com/package/http-errors?

File: test/client.js (hunk 1 above)
Comment (kibertoad) [reply test/client.js:338]: Any ideas? What aspects would you like emphasized? Axios uses `validateStatus` for similar config, got uses `throwHttpErrors`.

**Comment by kibertoad** [comment]
@ronag Hopefully everyone will win after this feature is properly refined. Can you elaborate on what changes would you like to be made on error structure? I followed structure of other Undici errors, should this one be different for some reason?

**Comment by kibertoad** [comment]
@ronag I gave it a shot, please let me know if revised error structure is what you wanted.

File: types/dispatcher.d.ts (hunk 2 above)
Comment (kibertoad) [reply types/dispatcher.d.ts:61]: done

File: lib/api/api-request.js [hunk 3]
```diff
@@ -89,6 +91,13 @@ class RequestHandler extends AsyncResource {
     const headers = this.responseHeaders === 'raw' ? util.parseRawHeaders(rawHeaders) : util.parseHeaders(rawHeaders)
 
     if (callback !== null) {
+      if (this.throwOnError && statusCode > 399) {
```
Comment (ronag) [inline lib/api/api-request.js:94]:
```suggestion
      if (this.throwOnError && statusCode >= 400) {
```

File: lib/core/request.js [hunk 4]
```diff
@@ -71,6 +72,8 @@ class Request {
 
     this.bodyTimeout = bodyTimeout
 
+    this.throwOnError = throwOnError !== undefined ? throwOnError : false
```
Comment (ronag) [inline lib/core/request.js:75]:
```suggestion
    this.throwOnError = throwOnError === true
```

**Review by ronag — APPROVED** [review]

File: test/client.js (hunk 1 above)
Comment (kibertoad) [reply test/client.js:338]: @ronag any preference on the name?

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by kibertoad** [comment]
@mcollina What is the release cadence? Any estimates on when new version can be published?

**Comment by kibertoad** [comment]
I think some flakiness for Windows/Node 16 build was introduced again with this PR. I'll try to fiddle with the new tests added here, helped in the past.

**Comment by mcollina** [comment]
thanks, once that's stable I'll ship a release.

**Comment by s97712** [comment]
This will lose the error stack.

**Comment by kibertoad** [comment]
@s97712 can you elaborate? there is no error stack on error code replies per se

**Comment by s97712** [comment]
If I use throwOnError to throw an error, I will get the following stack trace. 
This will be difficult to pinpoint the exact location of the error.
```bash
  ResponseStatusCodeError: Bad Request

  - util.js:36 getResolveErrorBodyCallback
    [admin]/[undici@5.23.0]/[undici]/lib/api/util.js:36:34

  - task_queues:95 processTicksAndRejections
    node:internal/process/task_queues:95:5

```

**Pull request merged**